runtime/connectors/blob/parquetreader.go

```diff
@@ -0,0 +1,146 @@
+	// reader to convert parquet objects to arrow objects
+	fileReader, err := pqarrow.NewFileReader(pf, arrowReadProperties, mem)
+	if err != nil {
+		return err
+	}
+
+	numRowGroups := pf.NumRowGroups()
+	if numRowGroups == 0 {
+		return fmt.Errorf("invalid parquet")
+	}
+
+	records, err := estimateRecords(ctx, pf, fileReader, option)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		for _, rec := range records {
+			rec.Release()
+		}
+	}()
+
+	schema, err := fileReader.Schema()
+	if err != nil {
+		return err
+	}
+
+	table := array.NewTableFromRecords(schema, records)
+	defer table.Release()
+
+	// duck db requires root Repetitions to be required
+	// keeping compressions as uncompressed since file will be immediately consumed and deleted
+	wp := parquet.NewWriterProperties(
+		parquet.WithRootRepetition(parquet.Repetitions.Required),
+		parquet.WithCompression(compress.Codecs.Uncompressed),
+		parquet.WithAllocator(mem))
+	return pqarrow.WriteTable(
+		table,
+		fw,
+		table.NumRows(),
+		wp,
+		pqarrow.NewArrowWriterProperties(pqarrow.WithAllocator(mem)),
+	)
+}
+
+func containerForRecordLimiting(strategy runtimev1.Source_ExtractPolicy_Strategy, limit int) (container.Container[arrow.Record], error) {
+	switch strategy {
+	case runtimev1.Source_ExtractPolicy_STRATEGY_TAIL:
+		return container.NewTailContainer(limit, func(rec arrow.Record) { rec.Release() })
+	case runtimev1.Source_ExtractPolicy_STRATEGY_HEAD:
+		return container.NewBoundedContainer[arrow.Record](limit)
+	default:
+		// No option selected - this should not be used for partial downloads though
+		return container.NewUnboundedContainer[arrow.Record]()
+	}
+}
+
+// estimateRecords estimates the number of rows to fetch based on extract policy
+// each arrow.Record will hold batchSize number of rows
+func estimateRecords(ctx context.Context, reader *file.Reader, pqToArrowReader *pqarrow.FileReader, config *extractOption) ([]arrow.Record, error) {
+	rowIndexes := arrayutil.RangeInt(0, reader.NumRowGroups(), config.strategy == runtimev1.Source_ExtractPolicy_STRATEGY_TAIL)
+
+	var (
+		// row group indices that we need
+		reqRowIndices []int
+		cumSize       uint64
+		rows          int64
+	)
+	for _, index := range rowIndexes {
+		reqRowIndices = append(reqRowIndices, index)
+		rowGroup := reader.RowGroup(index)
+		// we fetch compressed size from fs instead of ByteSize(uncompressed size)
+		rowGroupSize := rowGroup.MetaData().TotalCompressedSize()
+		rowCount := rowGroup.NumRows()
+
+		if cumSize+uint64(rowGroupSize) > config.limitInBytes {
+			// taking entire rowgroup crosses allowed size
+			perRowSize := uint64(rowGroupSize / rowCount)
+			rows += int64((config.limitInBytes - cumSize) / perRowSize)
+			break
+		}
+		cumSize += uint64(rowGroupSize)
+		rows += rowCount
+	}
+
+	r, err := pqToArrowReader.GetRecordReader(ctx, arrayutil.RangeInt(0, reader.RowGroup(0).NumColumns(), false), reqRowIndices)
```

Comment:
I am not sure about whether different row groups can have different NumColumns. But since there is a separate `NumColumns` field in each row group it may not be safe to assume that it will be always same. 
But we can solve it here at least by passing nil in which case the lib will initialise the list of columns from the schema.